cmd/pinniped-proxy/src/pinniped.rs

```diff
@@ -131,7 +133,17 @@ async fn call_pinniped_exchange(authorization: &str, k8s_api_server_url: &str, k
         Some(a) => a.to_string(),
         None => authorization.to_string(),
     };
-    let token_creds: Api<TokenCredentialRequest> = Api::namespaced(client.clone(), &pinniped_namespace);
+    
+    let pinniped_version = env::var(PINNIPED_VERSION)?;
+    let token_creds: Api<TokenCredentialRequest>;
+    
+    // Pinniped >= 0.6.0 changed the APIs to be global instead of namespaced
+    if Version::parse(&pinniped_version) >  Version::parse("0.5.0") {
+        token_creds = Api::all(client.clone());
+    }else{
+        token_creds = Api::namespaced(client.clone(), &pinniped_namespace);
+    }
```

Comment:
If we do go ahead with this change and so it needs testing, I'd recommend just pulling it the functionality that you're adding and just unit-testing that:
```
fn get_token_credential_request() -> ...
```
if there's fields on the result which can be used, as I think I'm only testing the failure cases for this function atm (haven't looked for the equivalent of `httptest` for staging test requests.

Reply:
I have extracted the logic to a function, but I'm unable to test it properly. I was thinking in something like:

```rust
let actual_tcr = get_token_credential_request(client, pinniped_namespace);
let expected_tcr: Api<TokenCredentialRequest> = Api::all(client.clone());

assert_eq!(actual_tcr, expected_tcr);
```

However, it complains about something related to tokio and I had to introduce `#[cfg(feature = "rt-threaded")]` to pass the test. But, honestly, I dunno what is doing :S

Reply:
So this is only required because in the test you're calling `Client::try_default().await?;` which is an async function (which you're waiting for). I did a bit of a refactor so that you can create a client the same way the code above does (which isn't async), with:

```diff
--- a/cmd/pinniped-proxy/src/pinniped.rs
+++ b/cmd/pinniped-proxy/src/pinniped.rs
@@ -120,24 +120,28 @@ pub struct ClusterCredential {
     client_key_data: String,
 }
 
-/// call_pinniped_exchange returns the resulting TokenCredentialRequest with Status after requesting a token credential exchange.
-async fn call_pinniped_exchange(authorization: &str, k8s_api_server_url: &str, k8s_api_ca_cert_data: &[u8]) -> Result<TokenCredentialRequest> {
-    let pinniped_namespace = env::var(DEFAULT_PINNIPED_NAMESPACE)?;
-
+fn get_client_config(k8s_api_server_url: &str, k8s_api_ca_cert_data: &[u8], pinniped_namespace: String) -> Result<kube::Client> {
     let mut config = Config::new(Url::parse(k8s_api_server_url).context("Failed parsing url for exchange")?);
     config.default_ns = pinniped_namespace.clone();
     let x509 = X509::from_pem(k8s_api_ca_cert_data).context("error creating x509 from pem")?;
     let der = x509.to_der().context("error creating der from x509")?;
     config.root_cert = Some(vec!(der));
-    let client = Client::new(Service::try_from(config)?);
+
+    Ok(Client::new(Service::try_from(config)?))
+}
+
+/// call_pinniped_exchange returns the resulting TokenCredentialRequest with Status after requesting a token credential exchange.
+async fn call_pinniped_exchange(authorization: &str, k8s_api_server_url: &str, k8s_api_ca_cert_data: &[u8]) -> Result<TokenCredentialRequest> {
+    let pinniped_namespace = env::var(DEFAULT_PINNIPED_NAMESPACE)?;
+    let client = get_client_config(k8s_api_server_url, k8s_api_ca_cert_data, pinniped_namespace.clone())?;
 
     let auth_token = match authorization.to_string().strip_prefix("Bearer ") {
         Some(a) => a.to_string(),
         None => authorization.to_string(),
     };
```

With that I could update the test so it wasn't async at all, but unfortunely (for us) the item that we'd need to test in the client is `client.resource` which is a private member on client (sorry, hadn't checked that). The best way to test this, like any other language, would be to use a fixture which starts up a fake api server and records the URLS, but that's not something for this PR or something we should do now, imo. I'd just remove this test.